### ISSUE TITLE
Parsing output of yolo segmentation tensors

### ISSUE THREAD

**Issue by nick-ala** [ultralytics/ultralytics#17381]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

*I would like to start by saying that if this is a duplicate issue I apologize, I am asking because I have not been able to find any similar questions with a quick search. Now, on to the questions.*
### Background: 
The overall goal is to train a custom ultralytics yolov9 model for segmentation, export the model as a torchscript file for use in cpp production, run the model within my cpp code base, and parse the results. On the python side of things, I have it all working to my knowledge. I wont include the entire script I wrote, but after training I export the model as such:
`model.export(format="torchscript")`
This gets me to the cpp and parsing side of things, which is where my questions lie. Right now, I have a front to back solution, but the results are not as good as expected. The results from my cpp code are far less accurate and detailed than the results from the same model exported as a .pt file and ran in a python environment. It seems to be in the way I am interpreting output tensors, so I was hoping to walk through how I do this, and see if anyone has any insights as to what I might be doing wrong. 
### Postprocessing
My model returns two tensors, lets call them output0 `{1, 40, 8400}` and output1 `{1, 160, 160, 32}`. By my understanding, output0 should be `{batch_size, 4 + num_classes + num_masks, num_candidate_detections}`. This checks out because my batch size is 1, I have 4 classes, and the number of masks should be 32 for yolov9. output1 also adds up for being mask data. With that being said, here is how I have gone about extracting information from these tensors:

I start by parsing outputs in the following way:
- permute output0 into a `detections` tensor by making it contiguous, permuting it to `{1, 8400, 40}`, then squeezing it to `{8400, 40}`
- Extract the bounding box information from the first 4 of 40 values for each of the 8400 detections. 
- Extract class logits from the next 4 (I have 4 classes) of 40 values for each of the 8400 detections.
- Extract the mask_coeffs from the final 32 of 40 values for each of the 8400 detections.
- Extracting mask prototypes by squeezing output1 to `{160, 160, 32}`

Looking at all the info I pulled from output0, I filter by max class scores, apply a confidence threshold, and end up with a list of selected boxes that meet all criteria, then apply NMS to the selected boxes. 

Finally, I compute masks for each detection as such:

- Get the mask from the mask prototype like so: 
`torch::Tensor  mask  =  torch::matmul(mask_coeff, mask_prototype.flatten(1));`
`mask  =  torch::sigmoid(mask.reshape({ mask_prototype.size(1), mask_prototype.size(2) }));`
- Interpolate the mask back up to the original image size using nearest neighbor interpolation
- Create a binary mask by thresholding the interpolated mask
- Get the region of interest for the current mask, which is just whichever bounding box it is within
- Select just the portion of the mask within the ROI, then apply that to the image

### Tangible Questions
- Does exporting the model as a torchscript file directly as I showed in the background section produce the same result as tracing a pt model would? 
- If I export the same model as a pt file and as a torchscript file, should outputs from the two be the same? If not, how similar should their performances be?
- Are there any glaring errors in my post processing logic? I feel like I have to be interpreting the results wrong because this code produces results that are just so much worse than using the ultralytics libraries for predictions in python. Especially since the model weights show no signs of quantization, this is just baffling to me. 

### Closing thoughts
I understand this is a lot of information with a lot of questions. To anyone who has read this through, regardless of if you have an answer, thank you so much for your time. I understand that some points of this are a bit vague, and that's because this is for a work matter, so I do not want to share real blocks of code directly. If anyone has any clarifying questions, I will be very attentive to this post.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @nick-ala, thank you for reaching out and for your detailed question on the Ultralytics repository 🚀!

We appreciate the thorough context you've provided regarding your issue with parsing YOLO segmentation output tensors in C++. To best assist you:

- If this is a 🐛 Bug Report concerning discrepancies between Python and C++ results, please ensure to include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) that highlights the issue clearly. This will help us debug significantly.
  
- If it's a technical ❓ Question about segmentation post-processing or model export, ensure all relevant details are included to facilitate an accurate diagnosis.

🔍 In the meantime, you might find it helpful to revisit the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to cross-check whether any preprocessing or post-processing tweaks might enhance your results.

💡 For real-time help and discussions, join our community on [Discord](https://ultralytics.com/discord) 🎧 where many knowledgeable users and developers come together to solve queries interactively. You might also explore our [Discourse](https://community.ultralytics.com) for in-depth topic discussions, or engage with peers on our [Subreddit](https://reddit.com/r/ultralytics).

## Documentation & Setup

You should also make sure your setup is using the latest `ultralytics` package:

```bash
pip install -U ultralytics
```

Ensure your environment meets all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), with [**Python>=3.8**](https://www.python.org/) and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/).

## Verified Environments

YOLO running environments are listed below, pre-installed with all necessary dependencies, from which you can choose to optimize your workflow:

- **Notebooks** with free GPU access: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud and Virtual Environments**: Available on [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), and via a [Docker Image](https://hub.docker.com/r/ultralytics/ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a><br>If this badge is green, all our CI tests are passing, ensuring core functionality across diverse systems.

This is an automated response, but rest assured, an Ultralytics engineer will assist you soon. Thanks for your patience! 🚀

**Comment by Y-T-G** [comment]
You can follow this example for preprocessing and postprocessing.

https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python

**Comment by nick-ala** [comment]
@Y-T-G I have taken a look through the examples there, and the closest example I can see is this one:
https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-LibTorch-CPP-Inference/main.cc

It does everything in a very similar manner to my own code, and I was able to learn a few things about the letterbox resizing implementation. But I am still struggling with some of the mask decoding, and the example above is just bounding boxes, not segmentation. Do you have any idea of there are examples for instance segmentation? The pre and post processing will change a bit.
